### DISCUSSION TITLE
numerical issue biharmonic_coordinates

### DISCUSSION THREAD

**Discussion by imanolooo** [libigl/libigl#2052] [Q&A]
Hi,

I'm really newie using igl. I'm trying to use biharmonic_coordinates in 2D and, although I'm starting with a very simple case, I'm not successing at all... When I call to biharmonic_coordinates, it didn't works and I get a "Error: Numerical issue" that I don't understand...

First, I create the vertices and triangles representing a grid:

```
_vertices = Eigen::MatrixXd(width*height,2);
      auto index = 0;
      for(auto j = -height/2; j <= height/2; ++j) {
          for(auto i = -width/2; i <= width/2; ++i) {
              _vertices(index,0) = i;
              _vertices(index,1) = j;
              index++;
          }
      }


   _faces = Eigen::MatrixXi((width-1)*(height-1)*2,3);
    auto current = 0;
    index = 0;
    for(auto j = 0; j < height-1; ++j) {
        for(auto i = 0; i < width-1; ++i) {
            _faces(index,0) = current;
            _faces(index,1) = current + width;
            _faces(index,2) = current + 1;
            ++index;
            _faces(index,0) = current + 1;
            _faces(index,1) = current + width;
            _faces(index,2) = current + 1 + width;
            ++index;
            current++;
        }
        current++;
    }_vertices = Eigen::MatrixXd(width*height,2);
    auto index = 0;
    for(auto j = -height/2; j <= height/2; ++j) {
        for(auto i = -width/2; i <= width/2; ++i) {
            _vertices(index,0) = i;
            _vertices(index,1) = j;
            index++;
        }
    }

    _faces = Eigen::MatrixXi((width-1)*(height-1)*2,3);
    auto current = 0;
    index = 0;
    for(auto j = 0; j < height-1; ++j) {
        for(auto i = 0; i < width-1; ++i) {
            _faces(index,0) = current;
            _faces(index,1) = current + width;
            _faces(index,2) = current + 1;
            ++index;
            _faces(index,0) = current + 1;
            _faces(index,1) = current + width;
            _faces(index,2) = current + 1 + width;
            ++index;
            current++;
        }
        current++;
    }
```
Then, I select the position of 4 vertices in the V Eigen::MatrixXd and I try to compute the weights... 
```
std::vector<std::vector<int>> S;
    S.push_back({6});
    S.push_back({8});
    S.push_back({16});
    S.push_back({18});

    int k = 2;

    std::cout << igl::biharmonic_coordinates(_vertices,_faces,S,k,_weights) << std::endl;
```
I don't really see what can be wrong as the indexes of S are correct and I use triangles for T as it is written in biharmonic_coordinates.h file...

Thanks a lot for your help!

Imanol

**Comment by alecjacobson** [comment]
Could you please share your input data?

**Reply by imanolooo** [reply]
Hi,
thanks for your quik answer. The input data is generated on that loops. I'm using a regular grid of of 5x5 vertices centered on (0,0). The 25 vertices are:
```
-2 -2
-1 -2
 0 -2
 1 -2
 2 -2
-2 -1
-1 -1
 0 -1
 1 -1
 2 -1
-2  0
-1  0
 0  0
 1  0
 2  0
-2  1
-1  1
 0  1
 1  1
 2  1
-2  2
-1  2
 0  2
 1  2
 2  2
```
and the 32 triangles are
 ```
0  5  1
 1  5  6
 1  6  2
 2  6  7
 2  7  3
 3  7  8
 3  8  4
 4  8  9
 5 10  6
 6 10 11
 6 11  7
 7 11 12
 7 12  8
 8 12 13
 8 13  9
 9 13 14
10 15 11
11 15 16
11 16 12
12 16 17
12 17 13
13 17 18
13 18 14
14 18 19
15 20 16
16 20 21
16 21 17
17 21 22
17 22 18
18 22 23
18 23 19
19 23 24
```
I have chosen as control points the vertices 6, 8, 16 and 18 that corresponds to (-1,-1), (1,-1), (-1,1) and (1,1) respectively.

Imanol

**Reply by imanolooo** [reply]
Hi again,
I've just converted the vertices to 3D (z= 0) and the problem is solved. Now, igl can compute the biharmonic coordinates... Should the vertices be always in 3d?
thanks!

**Reply by imanolooo** [reply]
Hi again,

I've modified a bit the data and I'm getting numerical issues again... 

When I use the original 3D plane, everything works smooth:
![plane00](https://user-images.githubusercontent.com/17313498/176087760-13f40ad1-f0ca-4eb6-a656-71401afab735.png)

But, when I use a modified version to include the control points in the mesh, I'm getting numerical issues again...
![planeCP01](https://user-images.githubusercontent.com/17313498/176087814-3e7d88a6-fcbe-4520-8fe9-40a4c2144a97.png)

I don't see why it can happens... I have converted both to obj and both are well read by Meshlab, Blender, etc...

The S and k are configured in the same way in both. So I imagine the error can't be here...

I've debugged a bit the code and the error is launched in min_quad_with_fixed.impl.h in the line 228. I don't know very well how to debug it, or what to look for, but if i check the llt m_factorizationIsOk = true and m_analysisIsOk = true, but m_info = Eigen::NumericalIssue...

Any idea?

Thanks for your help and time!

Imanol
